File: src/cmds/http.ts
import { run } from '../core/runner';
import { stripAnsi, truncate, formatBytes } from '../core/utils';
import { deduplicateLines, extractStructure } from '../core/filter';
import { safeJsonParse } from '../core/utils';
import { HandlerResult } from './git';

// HTTP fetchers (curl/wget) and `env`. curl/wget bodies are only compressed when
// large (JSON → structure, otherwise dedup+truncate); `env` is redacted to keys
// only so secrets never reach the model.

export function handleCurl(args: string[], ultra: boolean): HandlerResult {
  const result = run('curl', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    filtered = summarizeBody(result.stdout || raw, ultra);
  } catch {
    filtered = raw;
  }
  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'curl',
    execMs: result.execMs,
  };
}

export function handleWget(args: string[], ultra: boolean): HandlerResult {
  const result = run('wget', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    const clean = stripAnsi(raw);
    // wget prints progress + a "saved [N]" line to stderr.
    const saved = /saved\s+\[(\d+)(?:\/\d+)?\]/.exec(clean);
    const status = /HTTP request sent.*?\s(\d{3})\b/.exec(clean);
    if (saved) filtered = ultra ? `✓${formatBytes(parseInt(saved[1], 10))}` : `✓ downloaded ${formatBytes(parseInt(saved[1], 10))}${status ? ` (HTTP ${status[1]})` : ''}`;
    else if (result.exitCode !== 0) filtered = clean.split('\n').filter((l) => /error|failed|unable/i.test(l)).slice(0, 5).join('\n') || truncate(clean, 8);
    else filtered = ultra ? '✓' : '✓ ok';
  } catch {
    filtered = raw;
  }
  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'wget',
    execMs: result.execMs,
  };
}

// `env` / `printenv` — show variable NAMES only. Values commonly hold tokens and
// secrets, so we never echo them to the model.
export function handleEnv(args: string[], ultra: boolean): HandlerResult {
  const bin = args[0] === '__printenv__' ? 'printenv' : 'env';
  const realArgs = args[0] === '__printenv__' ? args.slice(1) : args;
  const result = run(bin, realArgs);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    const keys = stripAnsi(result.stdout)
      .split('\n')
      .map((l) => /^([A-Za-z_][A-Za-z0-9_]*)=/.exec(l)?.[1])
      .filter((k): k is string => !!k)
      .sort();
    if (keys.length === 0) filtered = truncate(stripAnsi(raw).trim(), 10);
    else if (ultra) filtered = `${keys.length} vars`;
    else filtered = `${keys.length} environment variables (values redacted):\n${keys.join(', ')}`;
  } catch {
    filtered = raw;
  }
  return {
    filteredOutput: filtered || 'ok',
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'env',
    execMs: result.execMs,
  };
}

function summarizeBody(body: string, ultra: boolean): string {
  const clean = stripAnsi(body).trim();
  if (clean === '') return ultra ? '(empty)' : '(empty response)';

  // JSON → collapse to a key/type skeleton when large.
  const parsed = safeJsonParse(clean);
  if (parsed && typeof parsed === 'object') {
    const bytes = Buffer.byteLength(clean, 'utf8');
    if (bytes <= 800) return clean; // small JSON: pass through
    const structure = JSON.stringify(extractStructure(parsed), null, ultra ? 0 : 2);
    return `JSON response (${formatBytes(bytes)}), structure:\n${truncate(structure, ultra ? 20 : 60)}`;
  }

  // HTML → strip tags, note size.
  if (/^\s*<(!doctype|html)/i.test(clean)) {
    const text = clean.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
    const title = /<title>([^<]+)<\/title>/i.exec(clean)?.[1]?.trim();
    return `HTML (${formatBytes(Buffer.byteLength(clean, 'utf8'))})${title ? ` — "${title}"` : ''}: ${truncate(text, ultra ? 3 : 8)}`;
  }

  const lines = clean.split('\n');
  if (lines.length <= (ultra ? 8 : 40)) return clean;
  return truncate(deduplicateLines(clean), ultra ? 8 : 40);
}

File: internal/handlers/http.go
package handlers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/filter"
	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

// HTTP fetchers (curl/wget) and env. curl/wget bodies are only compressed when
// large (JSON -> structure, otherwise dedup+truncate); env is redacted to keys
// only so secrets never reach the model.

func HandleCurl(args []string, ultra bool) Result {
	r := run.Run("curl", args)
	raw := combined(r)
	filtered := summarizeBody(firstNonEmpty(r.Stdout, raw), ultra)
	return finalizeRaw(filtered, r, raw, "curl")
}

var (
	wgetSavedRe  = regexp.MustCompile(`saved\s+\[(\d+)(?:/\d+)?\]`)
	wgetStatusRe = regexp.MustCompile(`HTTP request sent.*?\s(\d{3})\b`)
	wgetErrRe    = regexp.MustCompile(`(?i)error|failed|unable`)
	envVarRe     = regexp.MustCompile(`^([A-Za-z_][A-Za-z0-9_]*)=`)
	htmlDocRe    = regexp.MustCompile(`(?i)^\s*<(!doctype|html)`)
	htmlTagRe    = regexp.MustCompile(`<[^>]+>`)
	whitespaceRe = regexp.MustCompile(`\s+`)
	htmlTitleRe  = regexp.MustCompile(`(?i)<title>([^<]+)</title>`)
)

func HandleWget(args []string, ultra bool) Result {
	r := run.Run("wget", args)
	raw := combined(r)
	clean := util.StripAnsi(raw)
	var filtered string

	switch {
	case wgetSavedRe.MatchString(clean):
		size := util.FormatBytes(atoi(wgetSavedRe.FindStringSubmatch(clean)[1]))
		if ultra {
			filtered = "✓" + size
		} else {
			filtered = "✓ downloaded " + size
			if s := wgetStatusRe.FindStringSubmatch(clean); s != nil {
				filtered += " (HTTP " + s[1] + ")"
			}
		}
	case r.ExitCode != 0:
		var errs []string
		for _, l := range strings.Split(clean, "\n") {
			if wgetErrRe.MatchString(l) {
				errs = append(errs, l)
			}
		}
		filtered = strings.Join(limit(errs, 5), "\n")
		if filtered == "" {
			filtered = util.Truncate(clean, 8)
		}
	default:
		if ultra {
			filtered = "✓"
		} else {
			filtered = "✓ ok"
		}
	}
	return finalizeRaw(filtered, r, raw, "wget")
}

// HandleEnv shows variable NAMES only. Values commonly hold tokens and secrets, so
// we never echo them to the model.
func HandleEnv(args []string, ultra bool) Result {
	bin := "env"
	realArgs := args
	if len(args) > 0 && args[0] == "__printenv__" {
		bin = "printenv"
		realArgs = args[1:]
	}
	r := run.Run(bin, realArgs)
	raw := combined(r)

	var keys []string
	for _, l := range strings.Split(util.StripAnsi(r.Stdout), "\n") {
		if m := envVarRe.FindStringSubmatch(l); m != nil {
			keys = append(keys, m[1])
		}
	}
	sort.Strings(keys)

	var filtered string
	switch {
	case len(keys) == 0:
		filtered = util.Truncate(strings.TrimSpace(util.StripAnsi(raw)), 10)
	case ultra:
		filtered = fmt.Sprintf("%d vars", len(keys))
	default:
		filtered = fmt.Sprintf("%d environment variables (values redacted):\n%s", len(keys), strings.Join(keys, ", "))
	}
	if filtered == "" {
		filtered = "ok"
	}
	return Result{Filtered: filtered, Exit: r.ExitCode, Raw: raw, CmdType: "env", ExecMs: r.ExecMs}
}

func summarizeBody(body string, ultra bool) string {
	clean := strings.TrimSpace(util.StripAnsi(body))
	if clean == "" {
		if ultra {
			return "(empty)"
		}
		return "(empty response)"
	}

	// JSON -> collapse to a key/type skeleton when large.
	var parsed any
	if json.Unmarshal([]byte(clean), &parsed) == nil {
		switch parsed.(type) {
		case map[string]any, []any:
			size := len(clean)
			if size <= 800 {
				return clean
			}
			structure := jsonStringify(filter.ExtractStructure(parsed), !ultra)
			return fmt.Sprintf("JSON response (%s), structure:\n%s",
				util.FormatBytes(size), util.Truncate(structure, ternInt(ultra, 20, 60)))
		}
	}

	// HTML -> strip tags, note size.
	if htmlDocRe.MatchString(clean) {
		text := strings.TrimSpace(whitespaceRe.ReplaceAllString(htmlTagRe.ReplaceAllString(clean, " "), " "))
		head := fmt.Sprintf("HTML (%s)", util.FormatBytes(len(clean)))
		if t := htmlTitleRe.FindStringSubmatch(clean); t != nil {
			head += fmt.Sprintf(` - "%s"`, strings.TrimSpace(t[1]))
		}
		return fmt.Sprintf("%s: %s", head, util.Truncate(text, ternInt(ultra, 3, 8)))
	}

	lines := strings.Split(clean, "\n")
	if len(lines) <= ternInt(ultra, 8, 40) {
		return clean
	}
	return util.Truncate(filter.DeduplicateLines(clean), ternInt(ultra, 8, 40))
}

// jsonStringify mirrors JSON.stringify: no HTML escaping, optional 2-space indent,
// and no trailing newline.
func jsonStringify(v any, indent bool) string {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if indent {
		enc.SetIndent("", "  ")
	}
	if err := enc.Encode(v); err != nil {
		return ""
	}
	return strings.TrimRight(buf.String(), "\n")
}
